cmd/tmpfs/tmpfs.go

```diff
@@ -0,0 +1,347 @@
+	}
+	return msize, version, nil
+}
+
+// TODO handle aname?
+func (fs *fileServer) Rattach(fid protocol.FID, afid protocol.FID, uname string, aname string) (protocol.QID, error) {
+	if afid != protocol.NOFID {
+		return protocol.QID{}, fmt.Errorf("We don't do auth attach")
+	}
+
+	// There should be no .. or other such junk in the Aname. Clean it up anyway.
+	/*aname = path.Join("/", aname)
+	st, err := os.Stat(aname)
+	if err != nil {
+		return protocol.QID{}, err
+	}*/
+
+	root := fs.archive.Root()
+	fs.filesMutex.Lock()
+	fs.files[fid] = newFidEntry(root, uname)
+	fs.filesMutex.Unlock()
+
+	return root.Qid(), nil
+}
+
+func (fs *fileServer) Rflush(o protocol.Tag) error {
+	return nil
+}
+
+func (fs *fileServer) Rwalk(fid protocol.FID, newfid protocol.FID, paths []string) ([]protocol.QID, error) {
+	// Lookup the parent fid
+	parentEntry, err := fs.getFile(fid)
+	if err != nil {
+		return nil, fmt.Errorf("does not exist")
+	}
+
+	if len(paths) == 0 {
+		// Clone fid - point to same entry
+		fs.filesMutex.Lock()
+		defer fs.filesMutex.Unlock()
+		_, ok := fs.files[newfid]
+		if ok {
+			return nil, fmt.Errorf("FID in use: clone walk, fid %d newfid %d", fid, newfid)
+		}
+
+		fs.files[newfid] = newFidEntry(parentEntry.Entry, parentEntry.uname)
+		return []protocol.QID{}, nil
+	}
+
+	walkQids := make([]protocol.QID, len(paths))
+
+	var i int
+	var pathcmp string
+	currEntry := parentEntry.Entry
+	for i, pathcmp = range paths {
+		var ok bool
+		var dir *tmpfs.Directory
+		if dir, ok = currEntry.(*tmpfs.Directory); ok {
+			if pathcmp == ".." {
+				currEntry = dir.Parent()
+			} else {
+				currEntry, ok = dir.ChildByName(pathcmp)
+				if !ok {
+					// From the RFC: If the first element cannot be walked for any
+					// reason, Rerror is returned. Otherwise, the walk will return an
+					// Rwalk message containing nwqid qids corresponding, in order, to
+					// the files that are visited by the nwqid successful elementwise
+					// walks; nwqid is therefore either nwname or the index of the
+					// first elementwise walk that failed. The value of nwqid cannot be
+					// zero unless nwname is zero. Also, nwqid will always be less than
+					// or equal to nwname. Only if it is equal, however, will newfid be
+					// affected, in which case newfid will represent the file reached
+					// by the final elementwise walk requested in the message.
+					//
+					// to sum up: if any walks have succeeded, you return the QIDS for
+					// one more than the last successful walk
+					if i == 0 {
+						return nil, fmt.Errorf("file does not exist")
+					}
+					// we only get here if i is > 0 and less than nwname,
+					// so the i should be safe.
+					return walkQids[:i], nil
+				}
+			}
+		}
+		walkQids[i] = currEntry.Qid()
+	}
+
+	fs.filesMutex.Lock()
+	defer fs.filesMutex.Unlock()
+	// this is quite unlikely, which is why we don't bother checking for it first.
+	if fid != newfid {
```

Comment:
?? it's very common. Is there a *measured* basis for this claim. B/c/ from watching a lot of 9p traffic I see this case all the time. I mean, every walk from the root starts with a clone to a new fid, as does almost every directory open I've seen. Maybe I missed something. But ... "quite unlikely" ... not a number.

Reply:
This possibly came from you - it was lifted from the ufs code 😁 Needs to be looked into though.

Reply:
then I'm full of poop. I have NO idea why I said this. Please beat up the author.